### PR TITLE
Hide Media source API only for background contents.

### PR DESCRIPTION
**Pull request by sangwoo108**
Resolves https://github.com/brave/brave-browser/issues/25765  - Youtube resolution is limited
Resolves https://github.com/brave/brave-browser/issues/25130  - streaming contents are unusable

And also as a side effect, 
Resolves https://github.com/brave/brave-browser/issues/25338  - thumbnail isn't correct(youtube implementation dependent, and not occur when using background contents)

Reportedly, media source API also affects video quality. Now that we have critical downside, We shouldn't hide the API for tabs' web contents.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by sangwoo108** [comment]
Hi, @simonhong @goodov ! You may remember that we've talked how to deal with Media Source API. But now things are changed. Not only streaming contents, we have downside for normal video contents on Youtube. When hiding the media source API, videos resolution is limitied to low quality. I think we shouldn't disable the API for tabs contents in terms of web compatibility.

So, I made this PR to disable Media source API only for background web contents - which users don't interact with. And when we try downloading videos from youtube, it'll always use the background contents. Please take a look :)


p.s. @simonhong , thanks to your previous work on cosemetic filter, I could do this a way easier. thanks.

File: components/playlist/playlist_download_request_manager.h [hunk 1]
```diff
@@ -71,12 +71,16 @@ class PlaylistDownloadRequestManager
...
 
   // Update |web_prefs| if we want for |web_contents|.
-  void ConfigureWebPrefsforBackgroundWebContents(
+  void ConfigureWebPrefsForBackgroundWebContents(
       content::WebContents* web_contents,
       blink::web_pref::WebPreferences* web_prefs);
 
   content::WebContents* GetBackgroundWebContentsForTesting();
 
+  MediaDetectorComponentManager* media_detector_component_manager() {
```
Comment (goodov) [inline components/playlist/playlist_download_request_manager.h:80]: `const`?

File: components/playlist/renderer/playlist_render_frame_observer.cc [hunk 2]
```diff
@@ -24,17 +24,14 @@ PlaylistRenderFrameObserver::PlaylistRenderFrameObserver(
...
-      origins{
-          {blink::WebSecurityOrigin::Create(GURL("https://www.youtube.com"))}};
-
   const auto current_origin =
       render_frame()->GetWebFrame()->GetSecurityOrigin();
-  if (base::ranges::any_of(*origins, [&current_origin](const auto& origin) {
-        return origin.IsSameOriginWith(current_origin);
-      })) {
+  if (base::ranges::any_of(
+          render_frame()->GetBlinkPreferences().urls_to_hide_media_src_api,
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/playlist/renderer/playlist_render_frame_observer.cc]: Oh, I didn't know that there's a such a thing. Thanks!

**Comment by sangwoo108** [comment]
@goodov Thanks for comments :) I addressed them. PTAL again

File: chromium_src/third_party/blink/public/mojom/webpreferences/web_preferences.mojom [hunk 3]
```diff
@@ -5,7 +5,10 @@
 
 module blink.mojom;
 
+import "services/network/public/mojom/schemeful_site.mojom";
+
 [BraveExtend]
 struct WebPreferences {
   bool force_cosmetic_filtering;
+  array<network.mojom.SchemefulSite> urls_to_hide_media_src_api;
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply chromium_src/third_party/blink/public/mojom/webpreferences/web_preferences.mojom:13]: I'm on the fence. I think either way is fine. @goodov What do you think about this?

[the remark this answers is not in the document]
Comment (goodov) [reply chromium_src/third_party/blink/public/mojom/webpreferences/web_preferences.mojom:13]: I guess if we know the final url in `ConfigureWebPrefsForBackgroundWebContents`, then we can simplify everything to a `bool`, but I'm not sure how exactly WebPreferences work (at what point, etc.). Feel free to try, if you see this does work fine, then let's do it the simpler way.

**Review by simonhong — APPROVED** [review]
++ 👍

File: components/playlist/media_detector_component_manager.cc [hunk 4]
```diff
@@ -241,4 +246,17 @@ void MediaDetectorComponentManager::SetUseLocalScriptForTesting() {
   OnGetScript(kScript);
 }
 
+bool MediaDetectorComponentManager::ShouldHideMediaSrcAPI(
+    const GURL& url) const {
+  return base::ranges::any_of(sites_to_hide_media_src_api_,
+                              [&url](const auto& site_to_hide) {
+                                return site_to_hide == net::SchemefulSite(url);
```
Comment (goodov) [inline components/playlist/media_detector_component_manager.cc:253]: `net::SchemefulSite(url)` is created each time this lambda is called. pls move it from the lambda.

File: components/playlist/playlist_download_request_manager.cc [hunk 5]
```diff
@@ -256,11 +256,12 @@ void PlaylistDownloadRequestManager::DestroyWebContents() {
   web_contents_.reset();
 }
 
-void PlaylistDownloadRequestManager::ConfigureWebPrefsforBackgroundWebContents(
+void PlaylistDownloadRequestManager::ConfigureWebPrefsForBackgroundWebContents(
     content::WebContents* web_contents,
     blink::web_pref::WebPreferences* web_prefs) {
   if (web_contents_ && web_contents_.get() == web_contents) {
     web_prefs->force_cosmetic_filtering = true;
+    web_prefs->force_to_hide_media_src_api = true;
```
Comment (goodov) [inline components/playlist/playlist_download_request_manager.cc:264]: maybe just `hide_media_src_api`?

File: chromium_src/third_party/blink/public/mojom/webpreferences/web_preferences.mojom [hunk 6]
```diff
@@ -8,4 +8,5 @@ module blink.mojom;
 [BraveExtend]
 struct WebPreferences {
   bool force_cosmetic_filtering;
+  bool force_to_hide_media_src_api;
```
Comment (goodov) [inline chromium_src/third_party/blink/public/mojom/webpreferences/web_preferences.mojom:11]: can we add comments here for both members please?

**Review by goodov — APPROVED** [review]
lgtm with few nits.

File: components/playlist/playlist_download_request_manager.cc (hunk 5 above)
Comment (sangwoo108) [reply components/playlist/playlist_download_request_manager.cc]: Good. Done.

File: components/playlist/media_detector_component_manager.cc (hunk 4 above)
Comment (sangwoo108) [reply components/playlist/media_detector_component_manager.cc]: Good point! done.

File: chromium_src/third_party/blink/public/mojom/webpreferences/web_preferences.mojom (hunk 6 above)
Comment (sangwoo108) [reply chromium_src/third_party/blink/public/mojom/webpreferences/web_preferences.mojom]: Yep. done.

**Pull request merged**